sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/InvokeExprMeta.scala

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "400"}
+spark-rapids-shim-json-lines ***/
+
+package org.apache.spark.sql.rapids.shims
+
+import com.nvidia.spark.rapids.{DataFromReplacementRule, ExprMeta, GpuExpression, GpuOverrides, RapidsConf, RapidsMeta}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal, StructsToJson}
+import org.apache.spark.sql.catalyst.expressions.json.StructsToJsonEvaluator
+import org.apache.spark.sql.catalyst.expressions.objects.Invoke
+import org.apache.spark.sql.rapids.GpuStructsToJson
+import org.apache.spark.sql.types._
+
+class InvokeExprMeta(
+    invoke: Invoke,
+    conf: RapidsConf,
+    p: Option[RapidsMeta[_, _, _]],
+    r: DataFromReplacementRule) extends ExprMeta[Invoke](invoke, conf, p, r) {
+
+  /**
+   * Invoke is a dynamic expression, it can have different children.
+   */
+  override def isChildExprsCountDynamic: Boolean = true
+
+  override final def tagExprForGpu(): Unit = {
+    invoke match {
+      case Invoke(
+      Literal(eval: StructsToJsonEvaluator, _: ObjectType),
+      functionName: String,
+      _: StringType,
+      arguments: Seq[Expression],
+      methodInputTypes: Seq[AbstractDataType],
+      _: Boolean,
+      _: Boolean,
+      _: Boolean) if (functionName == "evaluate"
+        && arguments.size == 1
+        && methodInputTypes.size == 1) =>
+      // StructsToJson is equivalent to
+      // invoke(literal(StructsToJsonEvaluator), "evaluate", string_type, arguments, ...)
+      // here forward the checking to `StructsToJson`, in this way we can avoid duplicated checks
+      // definition
+      val child = invoke.arguments(0)
+      val stj = StructsToJson(eval.options, child, eval.timeZoneId)
+      val stjMeta = GpuOverrides.wrapExpr(stj, conf, None)
+        stjMeta.initReasons()
+        stjMeta.tagForGpu()
+      if(!stjMeta.canThisBeReplaced) {
+        val sb = new StringBuilder()
+        stjMeta.print(sb, depth = 0, all = false)
+        willNotWorkOnGpu(sb.toString())
+      }
+      case _ =>
+        // Unsupported invoke expr
+        willNotWorkOnGpu(s"Unsupported invoke expr: $invoke")
+    }
+  }
+
+  override def convertToGpu(): GpuExpression = {
+    invoke match {
+      case Invoke(
+      Literal(evaluator: StructsToJsonEvaluator, _: ObjectType),
+      functionName: String,
+      _: StringType,
+      arguments: Seq[Expression],
+      methodInputTypes: Seq[AbstractDataType],
+      _: Boolean,
+      _: Boolean,
+      _: Boolean) if (functionName == "evaluate"
+        && arguments.size == 1
+        && methodInputTypes.size == 1) =>
+        // Supported invoke expr which wraps an StructsToJsonEvaluator
+        val child = childExprs(1).convertToGpu().asInstanceOf[Expression]
+        GpuStructsToJson(evaluator.options, child, evaluator.timeZoneId)
```

Comment:
We can define an enum (or int) and specify it to a value that can be used here to know if a `StructsToJson` is matched. Doing this can avoid this complicatd duplicate match check.

```
supportedTarget match {
  case STRUCTS_TO_JSON =>
    val evaluator = invoke.targetObject.eval(null).asInstanceOf[StructsToJsonEvaluator]
    val child = childExprs(1).convertToGpu().asInstanceOf[Expression]
    GpuStructsToJson(evaluator.options, child, evaluator.timeZoneId)
  case _  =>
    throw new UnsupportedOperationException(s"Unsupported invoke expr: $invoke")
}
```